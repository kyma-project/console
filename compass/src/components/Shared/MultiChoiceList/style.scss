.multi-choice-list__list-element {
    list-style-type: none;
    display: flex;
    justify-content: space-between;
    vertical-align: middle;
    align-items: center;

    & > :first-child {
      margin-left: 3px;
    }
}

.multi-choice-list-modal__body {
  .fd-popover, .fd-popper__body {
    width: 100%;
  }
  
  .fd-dropdown {
    width: 100%;
  
    .fd-dropdown__control {
      display: flex;
      width: 100%;
      justify-content: space-between;
      justify-items: center;

      &::after {
        margin-top: -2px;
      }

      &::before {
        content: unset;
      }

    }
  }
}

.multi-choice-list__non-selected {
  list-style-type: none;
}