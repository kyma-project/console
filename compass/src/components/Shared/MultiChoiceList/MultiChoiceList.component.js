import React from 'react';
import PropTypes from 'prop-types';

import { Button, Dropdown, Icon } from '@kyma-project/react-components';
import { Menu } from 'fundamental-react';
import './style.scss';

MultiChoiceList.propTypes = {
  placeholder: PropTypes.string,
  updateItems: PropTypes.func.isRequired,
  currentlySelectedItems: PropTypes.array.isRequired,
  currentlyNonSelectedItems: PropTypes.array.isRequired,
  notSelectedMessage: PropTypes.string,
  noEntitiesAvailableMessage: PropTypes.string,
  displayPropertySelector: PropTypes.string,
};

MultiChoiceList.defaultProps = {
  placeholder: 'Choose items...',
  noEntitiesAvailableMessage: 'No more entries available',
};

export default function MultiChoiceList({
  placeholder,
  currentlySelectedItems,
  currentlyNonSelectedItems,
  updateItems,
  notSelectedMessage,
  noEntitiesAvailableMessage,
  displayPropertySelector,
}) {
  const [selectedItems, setSelectedItems] = React.useState(
    currentlySelectedItems,
  );
  const [nonSelectedItems, setNonselectedItems] = React.useState(
    currentlyNonSelectedItems,
  );

  function getDisplayName(item) {
    return displayPropertySelector ? item[displayPropertySelector] : item;
  }

  function selectItem(item) {
    const newSelectedItems = [...selectedItems, item];
    const newNonSelectedItems = nonSelectedItems.filter(i => i !== item);

    updateLists(newSelectedItems, newNonSelectedItems);
  }

  function unselectItem(item) {
    const newSelectedItems = selectedItems.filter(i => i !== item);
    const newNonSelectedItems = [...nonSelectedItems, item];

    updateLists(newSelectedItems, newNonSelectedItems);
  }

  function updateLists(newSelectedItems, newNonSelectedItems) {
    setSelectedItems(newSelectedItems);
    setNonselectedItems(newNonSelectedItems);

    updateItems(newSelectedItems, newNonSelectedItems);
  }

  function createAssignedEntitiesList() {
    const selectedEntitiesList = (
      <ul>
        {selectedItems.map(item => (
          <li
            className="multi-choice-list__list-element"
            key={getDisplayName(item)}
          >
            <span>{getDisplayName(item)}</span>
            <Button
              option="light"
              type="negative"
              onClick={() => unselectItem(item)}
            >
              <Icon size="l" glyph="decline" />
            </Button>
          </li>
        ))}
      </ul>
    );

    return selectedItems.length ? selectedEntitiesList : notSelectedMessage;
  }

  function createNonChoosenEntitiesDropdown() {
    const nonChoosenItemsList = nonSelectedItems.map(item => (
      <Menu.Item onClick={() => selectItem(item)}>
        {getDisplayName(item)}
      </Menu.Item>
    ));

    return (
      <Dropdown
        control={
          <Button dropdown>
            <span>
              {nonSelectedItems.length
                ? placeholder
                : noEntitiesAvailableMessage}
            </span>
          </Button>
        }
        placement="bottom"
      >
        {nonChoosenItemsList}
      </Dropdown>
    );
  }

  return (
    <section className="multi-choice-list-modal__body">
      {createAssignedEntitiesList()}
      {createNonChoosenEntitiesDropdown()}
    </section>
  );
}
