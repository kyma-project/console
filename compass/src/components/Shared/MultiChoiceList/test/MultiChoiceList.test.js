import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import * as mockData from './mockData';
import MultiChoiceList from '../MultiChoiceList.component';

describe('MultiChoiceList', () => {
  // for "Warning: componentWillMount has been renamed"
  console.error = jest.fn();
  console.warn = jest.fn();

  afterEach(() => {
    console.error.mockReset();
    console.warn.mockReset();
  });

  afterAll(() => {
    expect(console.error.mock.calls[0][0]).toMatchSnapshot();
    if (console.warn.mock.calls.length) {
      expect(console.warn.mock.calls[0][0]).toMatchSnapshot();
    }
  });

  it('Renders empty list with default caption props', () => {
    const component = renderer.create(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={[]}
        currentlyNonSelectedItems={[]}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders empty list with custom caption props', () => {
    const component = renderer.create(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={[]}
        currentlyNonSelectedItems={[]}
        placeholder="placeholder value"
        notSelectedMessage="not assigned message"
        noEntitiesAvailableMessage="no entities available message"
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders two lists of simple items', () => {
    console.error = jest.fn();

    const component = renderer.create(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={mockData.simpleSelected}
        currentlyNonSelectedItems={mockData.simpleNonselected}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();

    // catch "Warning: Each child in a list should have a unique \"key\" prop." comming from Fundamental
    expect(console.error.mock.calls.length).toBe(1);
    expect(console.error.mock.calls[0][0]).toMatchSnapshot();
  });

  it('Renders two lists of object items', () => {
    const component = renderer.create(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={mockData.selectedObjects}
        currentlyNonSelectedItems={mockData.nonSelectedObjects}
        displayPropertySelector="name"
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Calls updateItems when user clicks on selected items list', () => {
    const updateItems = jest.fn();

    const component = mount(
      <MultiChoiceList
        updateItems={updateItems}
        currentlySelectedItems={['a', 'b']}
        currentlyNonSelectedItems={[]}
      />,
    );

    const removeItemButton = component
      .find('.multi-choice-list__list-element')
      .filterWhere(n => n.text() === 'a')
      .find('button[data-test-id="unselect-button"]');

    removeItemButton.simulate('click');

    expect(updateItems).toHaveBeenCalledWith(['b'], ['a']);
  });

  it('Calls updateItems when user clicks on non selected items list', () => {
    const updateItems = jest.fn();

    const component = mount(
      <MultiChoiceList
        updateItems={updateItems}
        currentlySelectedItems={[]}
        currentlyNonSelectedItems={['b', 'a']}
      />,
    );

    // expand list
    const popoverControl = component.find('.fd-button.fd-dropdown__control');
    popoverControl.simulate('click');

    component.update();

    const addItemButton = component
      .find('span[data-test-id="select-button"]')
      .filterWhere(n => n.text() === 'a');

    addItemButton.simulate('click');

    expect(updateItems).toHaveBeenCalledWith(['a'], ['b']);
  });
});
