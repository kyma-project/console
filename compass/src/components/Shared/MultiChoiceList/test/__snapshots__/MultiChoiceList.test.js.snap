// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiChoiceList Renders empty list with custom caption props 1`] = `
<section
  className="multi-choice-list-modal__body"
>
  <p
    className="fd-has-font-style-italic"
  >
    not assigned message
  </p>
  <p
    className="fd-has-font-style-italic"
  >
    no entities available message
  </p>
</section>
`;

exports[`MultiChoiceList Renders empty list with default caption props 1`] = `
<section
  className="multi-choice-list-modal__body"
>
  <p
    className="fd-has-font-style-italic"
  />
  <p
    className="fd-has-font-style-italic"
  >
    No more entries available
  </p>
</section>
`;

exports[`MultiChoiceList Renders two lists of object items 1`] = `
<section
  className="multi-choice-list-modal__body"
>
  <ul>
    <li
      className="multi-choice-list__list-element"
    >
      <span>
        choosen 1
      </span>
      <button
        className="fd-button--light fd-button--negative"
        data-test-id="unselect-button"
        onClick={[Function]}
      >
        <span
          className="sap-icon--decline sap-icon--l"
        />
      </button>
    </li>
    <li
      className="multi-choice-list__list-element"
    >
      <span>
        choosen 2
      </span>
      <button
        className="fd-button--light fd-button--negative"
        data-test-id="unselect-button"
        onClick={[Function]}
      >
        <span
          className="sap-icon--decline sap-icon--l"
        />
      </button>
    </li>
  </ul>
  <div
    className="fd-dropdown"
  >
    <span
      onFocus={[Function]}
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="fd-popper__control fd-popover"
      >
        <button
          className="fd-button fd-dropdown__control"
          onClick={[Function]}
          type="button"
        >
          <span>
            Choose items...
          </span>
        </button>
      </div>
    </span>
  </div>
</section>
`;

exports[`MultiChoiceList Renders two lists of simple items 1`] = `
<section
  className="multi-choice-list-modal__body"
>
  <ul>
    <li
      className="multi-choice-list__list-element"
    >
      <span>
        choosen 1
      </span>
      <button
        className="fd-button--light fd-button--negative"
        data-test-id="unselect-button"
        onClick={[Function]}
      >
        <span
          className="sap-icon--decline sap-icon--l"
        />
      </button>
    </li>
    <li
      className="multi-choice-list__list-element"
    >
      <span>
        choosen 2
      </span>
      <button
        className="fd-button--light fd-button--negative"
        data-test-id="unselect-button"
        onClick={[Function]}
      >
        <span
          className="sap-icon--decline sap-icon--l"
        />
      </button>
    </li>
    <li
      className="multi-choice-list__list-element"
    >
      <span>
        choosen 3
      </span>
      <button
        className="fd-button--light fd-button--negative"
        data-test-id="unselect-button"
        onClick={[Function]}
      >
        <span
          className="sap-icon--decline sap-icon--l"
        />
      </button>
    </li>
  </ul>
  <div
    className="fd-dropdown"
  >
    <span
      onFocus={[Function]}
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="fd-popper__control fd-popover"
      >
        <button
          className="fd-button fd-dropdown__control"
          onClick={[Function]}
          type="button"
        >
          <span>
            Choose items...
          </span>
        </button>
      </div>
    </span>
  </div>
</section>
`;

exports[`MultiChoiceList Renders two lists of simple items 2`] = `"Warning: Each child in a list should have a unique \\"key\\" prop.%s%s See https://fb.me/react-warning-keys for more information.%s"`;
